overlord/snapstate/snapmgr.go

```diff
@@ -923,10 +923,16 @@ func (m *SnapManager) doLinkSnap(t *state.Task, _ *tomb.Tomb) error {
 
 	// if we just installed a core snap, request a restart
 	// so that we switch executing its snapd
-	if newInfo.Type == snap.TypeOS && release.OnClassic {
+	if release.OnClassic && newInfo.Type == snap.TypeOS {
 		t.Logf("Restarting snapd...")
 		st.Unlock()
-		st.RequestRestart()
+		st.RequestRestart(state.RestartDaemon)
+		st.Lock()
+	}
+	if !release.OnClassic && (newInfo.Type == snap.TypeOS || newInfo.Type == snap.TypeKernel) {
+		t.Logf("Restarting system...")
```

Comment:
Both of these messages feel a bit off.. should probably say something like:

`"Requested system restart."`
`"Requested daemon restart."`

As this is just a request, which will be respected in due time, and could even be canceled.